### PR TITLE
Renderer2D Update: Batch Renderer 2D.

### PR DESCRIPTION
**Pull request by TejasJD**
# Batch Renderer 2D
- Implemented Batch Renderer which highly optimizes the performance of the renderer by reducing the draw calls.
- It is recommended to test the Batch Renderer in Release Build for getting correct measures (Since Debug Build turns of all the compiler optimizations and adds some form of additional overhead beneficial for Debugging purposes but slows the overall performance).
- Added comments for documentation generation to the VertexArray.h, Buffer.h and Texture.h files.
## Test Hardware
- **Intel Core i7-9750H @ 2.60 GHz.**
- **16 GB RAM.**
- **6 GB VRAM NVIDIA GeForce GTX 1660 Ti.**
## Preliminary Test Results
[Texture](https://drive.google.com/file/d/1cHPtR2Cgt2GFC0O7DxEX-xIaleTcVnF7/view?usp=share_link) that was used during, the render tests with texture (Transparent Background).
image
- **Width x Height**: 2000 x 2000.
- **Format**: 32 bpp (bits per pixel) BGRA (Blue Green Red Alpha).
### Colored quadrilaterals without batch rendering.
image
- Framerate massively tanks as the total quadrilaterals rendered increased even by slightest amount.
- 1 draw call or batch = 1 quadrilateral.
### Colored quadrilaterals with batch rendering.
image
- Framerate remains above 60 even when the number of quads rendered are above 400k.
- 1 draw call or batch = 10000 quadrilaterals.
### Textured quadrilaterals without batch rendering.
image
- 1 draw call or batch = 1 quadrilateral.
### Textured quadrilaterals with batch rendering.
image
- 1 draw call or batch = 10000 quadrilaterals.

### CODE REVIEW

**Review by Minghui-Gao — APPROVED** [review]

File: HorizonEngine/src/HorizonEngine/Renderer/Buffer.h [hunk 1]
```diff
@@ -5,27 +5,35 @@
...
 		None = 0,
-		Vecf, Vec2f, Vec3f, Vec4f,
+		Float, Vec2f, Vec3f, Vec4f,
 		Mat3f, Mat4f, Mat3i, Mat4i,
-		Veci, Vec2i, Vec3i, Vec4i,
+		Int, Vec2i, Vec3i, Vec4i,
 		Bool
 	};
 
+	/// <summary>
```
Comment (alex-mulkerrins) [inline HorizonEngine/src/HorizonEngine/Renderer/Buffer.h:20]: Thanks for adding in documentation! makes it easier to understand

**Comment by alex-mulkerrins** [comment]
Are the main aspect of the batch rendering code in the renderer files of this PR or is all of it spread across the PR, I noticed we have some textures code in there as well, is that part of the batch rendering? Either way this branch works locally for me

**Review by alex-mulkerrins — APPROVED** [review]

**Comment by TejasJD** [comment]
> Are the main aspect of the batch rendering code in the renderer files of this PR or is all of it spread across the PR, I noticed we have some textures code in there as well, is that part of the batch rendering? Either way this branch works locally for me

It has some sections related to the Texture as well, because to batch render texture quadrilaterals, I had to update few things in the Texture as well. So, I changed a few things in the Texture API as well.

**Comment by alex-mulkerrins** [comment]
> > Are the main aspect of the batch rendering code in the renderer files of this PR or is all of it spread across the PR, I noticed we have some textures code in there as well, is that part of the batch rendering? Either way this branch works locally for me
> 
> It has some sections related to the Texture as well, because to batch render texture quadrilaterals, I had to update few things in the Texture as well. So, I changed a few things in the Texture API as well.

Oh ok sweet, I think I'm getting similar frame rates on my machine also for the coloured quadrilaterals, great work on this PR, this batch rendering should really improve the rendering power of horizon game engine

**Comment by prajakta-kulkarni-git** [comment]
Documentation helped me to understand. Thank you.

**Review by prajakta-kulkarni-git — APPROVED** [review]

**Pull request merged**